libs/auth/src/apdu.ts

```diff
@@ -262,24 +267,54 @@ export function constructTlv(
 export function parseTlv(
   tagAsByteOrBuffer: Byte | Buffer,
   tlv: Buffer
-): [Buffer, Buffer, Buffer] {
-  const tagLength = Buffer.isBuffer(tagAsByteOrBuffer)
-    ? tagAsByteOrBuffer.length
-    : 1;
+): [tag: Buffer, length: Buffer, value: Buffer] {
+  const expectedTagBytes = Buffer.isBuffer(tagAsByteOrBuffer)
+    ? tagAsByteOrBuffer
+    : Buffer.of(tagAsByteOrBuffer);
+  const expectedTagLength = expectedTagBytes.length;
+  const tagBytes = tlv.subarray(0, expectedTagLength);
+  assert(
+    tagBytes.equals(expectedTagBytes),
+    `TLV tag (${inspect(tagBytes)}) does not match expected tag (${inspect(
+      expectedTagBytes
+    )})`
+  );
 
-  let tlvLengthLength = 1; // Derpiest variable name, I know
-  const tlvLengthFirstByte = tlv.at(tagLength);
+  let tlvLengthLength: number;
+  let tlvLength: number;
+  const tlvLengthFirstByte = assertDefined(tlv.at(expectedTagLength));
+  const tlvLengthSecondByte = tlv.at(expectedTagLength + 1);
+  const tlvLengthThirdByte = tlv.at(expectedTagLength + 2);
```

Comment:
Slight preference for defining `tlvLengthSecondByte` and `tlvLengthThirdByte` within the `if` conditions for which they're relevant. Otherwise, what we store in `tlvLengthSecondByte` or `tlvLengthThirdByte` might actually be the start of the TLV value